lucille-core/src/main/java/com/kmwllc/lucille/stage/AddRandomField.java

```diff
@@ -0,0 +1,166 @@
+    this.fieldName = ConfigUtils.getOrDefault(config, "field_name", "data");
+    this.minNumOfTerms = ConfigUtils.getOrDefault(config, "min_num_of_terms", null);
+    this.maxNumOfTerms = ConfigUtils.getOrDefault(config, "max_num_of_terms", null);
+    this.isNested = ConfigUtils.getOrDefault(config, "is_nested", false);
+    this.dataArr = null;
+    this.cardinality = null;
+    this.uniqueValues = null;
+
+    if (this.minNumOfTerms == null ^ this.maxNumOfTerms == null) {
+      throw new StageException("Both minimum and maximum number of terms must be specified if either is specified");
+    }
+    if (this.minNumOfTerms == null && this.maxNumOfTerms == null) {
+      this.minNumOfTerms = 1;
+      this.maxNumOfTerms = 1;
+    }
+    if (this.minNumOfTerms > this.maxNumOfTerms) {
+      throw new StageException("Minimum number of terms must be less than or equal to maximum");
+    }
+  }
+
+  @Override
+  public void start() throws StageException {
+    this.dataArr = this.inputDataPath != null ? getFileData(this.inputDataPath) : null;
+    this.cardinality = ConfigUtils.getOrDefault(config, "cardinality",
+        this.dataArr != null ? this.dataArr.size() : this.maxNumOfTerms);
+    this.uniqueValues = getUniqueValues(this.dataArr != null, this.dataArr);
+
+    if (this.dataArr != null && this.cardinality > this.dataArr.size()) {
+      throw new StageException("Cardinality must be less than the number of lines in given file");
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    ArrayList<String> fieldDataArr = genFieldDataArr(uniqueValues, minNumOfTerms, maxNumOfTerms);
+    if (isNested) {
+      populateFieldsNested(fieldName, doc, fieldDataArr);
+    } else {
+      populateFieldsDefault(fieldName, doc, fieldDataArr);
+    }
+    return null;
+  }
+
+  private void populateFieldsNested(String fieldName, Document doc, ArrayList<String> fieldDataArr) {
+    ObjectMapper mapper = new ObjectMapper();
+    ArrayNode array = mapper.createArrayNode();
+    for (String value : fieldDataArr) {
+      array.add(mapper.createObjectNode().put("data", value));
+    }
+    doc.setField(fieldName, array);
+  }
+
+  private void populateFieldsDefault(String fieldName, Document doc, ArrayList<String> fieldDataArr) {
+    for (String value : fieldDataArr) {
+      doc.setOrAdd(fieldName, value);
+    }
+  }
+
+  private List<String> getFileData(String inputDataPath) throws StageException {
+    try {
+      List<String> data = Files.readAllLines(Path.of(inputDataPath));
+      return data;
+    } catch (IOException e) {
+      throw new StageException("Could not read provided file path", e);
+    }
+  }
+
+  private List<String> getUniqueValues(boolean dataExists, List<String> inputData) {
+    List<String> uniqueValues = null;
+    if (dataExists) {
+      List<String> initialData = new ArrayList<>(inputData);
```

Comment:
initialData might contain duplicates
if two identical values are chosen randomly and added to uniqueValuesSet, the size of that set only increases by 1
so at the end of the for loop, uniqueValuesSet could be smaller than the desired cardinality
therefore it might be better to start by putting initialData into a Set, then what is currently named uniqueValuesSet can just be a List

Reply:
Changed so that the initialData is passed into a set and then into an array so that all duplicates are removed